pandas/_libs/index.pyx

```diff
@@ -324,8 +336,19 @@ cdef class IndexEngine:
                         n_alloc += 10_000
                         result = np.resize(result, n_alloc)
 
-                    result[count] = j
-                    count += 1
+                    if val == -1:
```

Comment:
can you add comments explaining whats going on here, or at least pointing back to this PR/issue

Reply:
will do